layout: Implement `border-spacing` per CSS 2.1 § 17.6.1 and the legacy `cellspacing` attribute per HTML5 § 14.3.9.

pcwalton opened the pull request:
Table layout code has been refactored to push the spacing down to
rowgroups and rows; this will aid the implementation of
`border-collapse` as well.

r? @SimonSapin

pcwalton committed:
layout: Implement `border-spacing` per CSS 2.1 § 17.6.1 and the legacy
`cellspacing` attribute per HTML5 § 14.3.9.

Table layout code has been refactored to push the spacing down to
rowgroups and rows; this will aid the implementation of
`border-collapse` as well.

This commit also fixes two nasty issues in table layout:

* In fixed layout, extra space would not be divided among columns that
  had auto width but had nonzero minimum width.

* In automatic layout, extra space would be distributed to constrained
  columns as well even if unconstrained columns with percentage equal to
  zero were present.

pcwalton commented:
Rebased. r? @SimonSapin

pcwalton commented:
Rebased. r? @SimonSapin

SimonSapin commented:
Critic has 14 open issues on this PR, it looks like they haven’t been addressed. Rebasing in the middle of a review makes the rest of the review harder :(

pcwalton commented:
Addressed issues. The tests you asked me to add actually exposed some bugs in automatic and fixed table layout independent of spacing, so I fixed those too.

r? @SimonSapin

larsbergstrom commented:
The issues appear to have been addressed and the additional nasty issue fixes appear correct, to my understanding of the table layout code. Please squash and r+!
